tutorials/lab/rabbit.rst

```diff
@@ -0,0 +1,128 @@
+command line:
+
+.. code-block:: console
+
+	$ flux alloc -N2 --setattr=dw="#DW jobdw type=xfs capacity=10GiB name=project1"
+
+Requesting two XFS file systems in a batch script:
+
+.. code-block:: bash
+
+	#!/bin/bash
+
+	#FLUX: -N 2
+	#FLUX: -q pdebug
+	#FLUX: --setattr=dw="""
+	#FLUX: #DW jobdw type=xfs capacity=1TiB name=my_xfs
+	#FLUX: #DW jobdw type=lustre capacity=10GiB name=my_lustre
+	#FLUX: """
+
+	echo "Hello World!" > $DW_JOB_my_lustre/world.txt
+
+	flux submit -N2 -n2 /bin/bash -c "echo 'Hello World!' > $DW_JOB_my_xfs/world.txt"
+
+
+jobdw directive fields
+----------------------
+
+The **type** field can be one of ``xfs``, ``lustre``, ``gfs2``, or ``raw``.
+``lustre`` storage is shared by all nodes in the job. By contrast, the other types
+are for node-local storage. Processes on one node will not be able to read
+data written on other nodes. Currently only ``xfs`` and ``lustre`` are known
+to work properly.
+
+The **capacity** field describes how much storage to be allocated. For ``lustre``
+the capacity refers to the overall capacity. For all other types, the capacity refers
+to the capacity per node. See
+`this Wikipedia article <https://en.wikipedia.org/wiki/Byte#Multiple-byte_units>`_
+for the meaning of the suffixes.
+
+The **name** field determines the suffix to the ``DW_JOB_`` environment variable.
+See below for more detail.
+
+
+Using Rabbit Storage
+--------------------
+
+For each ``jobdw`` directive associated with your job, your job will have
+an environment variable ``DW_JOB_[name]`` where ``[name]`` is the value
+of the ``name`` field in the directive. The value of the environment variable
+will be the path to the associated file system.
+
+For instance, for a directive ``#DW jobdw type=xfs capacity=10GiB name=project1``,
+the associated job will have an environment variable ``DW_JOB_project1``.
+
+
+Data Movement Directives
+------------------------
+
+To request that files be moved to or from the rabbits, additional DW
+directives must be added to the job in addition to jobdw directives.
+The ``copy_in`` directive is for moving data to the rabbits before the job
+starts, and the ``copy_out`` directive is for moving data from the rabbits
+after the job completes.
+
+
+Data Movement Examples
+----------------------
+
+.. warning::
+
+	When writing ``copy_in`` and ``copy_out`` directives *on the command line*,
+	be careful to always escape the ``$`` character when writing ``DW_JOB_[name]``
+	variables. Otherwise your shell will expand them. This warning does not apply
+	to batch scripts.
+
+Requesting a 10 gigabyte XFS file system on the command line with data movement
+both to and from the rabbits:
+
+.. code-block:: console
+
+	$ flux alloc -N2 --setattr=dw="#DW jobdw type=xfs capacity=10GiB name=project1
+	#DW copy_in source=/p/lustre1/behlendo/zeros_in destination=\$DW_JOB_project1/
+	#DW copy_out source=\$DW_JOB_project1/ destination=/p/lustre1/behlendo/zeros_out/"
```

Comment:
How about this which is a little more generic.

```
$ flux alloc -N2 --setattr=dw="#DW jobdw type=xfs capacity=10GiB name=project1
#DW copy_in source=/p/lustre1/$USER/file_in destination=\$DW_JOB_project1/
#DW copy_out source=\$DW_JOB_project1/ destination=/p/lustre1/$USER/dir_out/"
```

While testing this out I also noticed I get the following error if the `/p/lustre1/$USER/file_in ` doesn't exist.  If it does things work as expected.

```
24.260s: job.exception type=exception severity=0 DWS/Rabbit interactions failed: workflow in 'Error' state too long: DW Directive 1: Staging operation failed
```

Reply:
Don't want everyone trying to copy your lustre data eh? I'll fix, and add a note that the source must exist.

Reply:
> and add a note that the source must exist.

Let's also open an issue for this.  This should probably be reported as a fatal error, not as a "it took too long so I gave up" kind of error.  As a user it's going to be very unclear what the problem was given that error.

Reply:
I'll open an issue. But the problem is more general: there's no way for Flux to know the difference between a transient and a fatal error. They're reported the same way. So no matter what the error is, Flux will report it as "it took too long so I gave up"

Reply:
Right, this is something HPEs software would need to return as a fatal error to Flux.  @bdevcich-hpe any thoughts on this?

Reply:
https://github.com/NearNodeFlash/NearNodeFlash.github.io/issues/46 and https://github.com/NearNodeFlash/NearNodeFlash.github.io/issues/47